opentelemetry-sdk/src/opentelemetry/sdk/metrics/view.py

```diff
@@ -0,0 +1,161 @@
+
+
+class ViewConfig:
+
+    UNGROUPED = 0
+    LABEL_KEYS = 1
+    DROP_ALL = 2
+
+
+class View:
+    def __init__(
+        self,
+        metric: InstrumentT,
+        aggregator: Aggregator,
+        label_keys: Sequence[str] = None,
+        config: ViewConfig = ViewConfig.UNGROUPED,
+    ):
+        self.metric = metric
+        self.aggregator = aggregator
+        if label_keys is None:
+            label_keys = []
+        self.label_keys = sorted(label_keys)
+        self.config = config
+
+    # Uniqueness is based on metric, aggregator type, ordered label keys and ViewConfig
+    def __hash__(self):
+        return hash(
+            (self.metric, self.aggregator, tuple(self.label_keys), self.config)
+        )
+
+    def __eq__(self, other):
+        return (
+            self.metric == other.metric
+            and self.aggregator.__class__ == other.aggregator.__class__
+            and self.label_keys == other.label_keys
+            and self.config == other.config
+        )
+
+
+class ViewManager:
+    def __init__(self):
+        self.views = defaultdict(set)  # Map[Metric, Set]
+        self._view_lock = threading.Lock()
+
+    def register_view(self, view):
+        with self._view_lock:
+            if view not in self.views[view.metric]:
+                self.views[view.metric].add(view)
+            else:
+                logger.warning("View already registered.")
+                return
+
+    def unregister_view(self, view):
+        with self._view_lock:
+            if self.views.get(view.metric) is None:
+                logger.warning("Metric for view does not exist.")
+            elif view in self.views.get(view.metric):
+                self.views.get(view.metric).remove(view)
+
+    def generate_view_datas(self, metric, labels):
+        view_datas = set()
+        views = self.views.get(metric)
+        # No views configured, use default aggregations
+        if views is None:
+            aggregator = get_default_aggregator(metric)
+            # Default config aggregates on all label keys
+            view_datas.add(ViewData(tuple(labels), aggregator))
+        else:
+            for view in views:
+                updated_labels = []
+                if view.config == ViewConfig.LABEL_KEYS:
+                    label_key_set = set(view.label_keys)
+                    for label in labels:
+                        # Only keep labels that are in configured label_keys
+                        if label[0] in label_key_set:
+                            updated_labels.append(label)
+                    updated_labels = tuple(updated_labels)
+                elif view.config == ViewConfig.UNGROUPED:
+                    updated_labels = labels
+                # ViewData that is duplicate (same labels and aggregator) will be
+                # aggregated together as one
+                view_datas.add(
+                    ViewData(tuple(updated_labels), view.aggregator)
+                )
+        return view_datas
```

Comment:
As I understand it, the goal here is fast aggregator updates. When the user creates a bound instrument, we grab any aggregators currently associated with that bound instrument's metric via a view. That way when the user calls `bound_instrument.record` we can update the aggregator immediately without another lookup.

I think I understand how you got here, but it's very surprising to see `BoundInstrument`s hold references to a bunch of `ViewData`s. I don't have a suggestion to improve this, but it looks very suspicious and I'm interested to hear what others think.

I think we would have come up with a very different design if we had written a single record path first, then added views, then added optimizations for bound instruments on top.

Reply:
I reworked the owner of ViewDatas to `View`, so bound instruments now only have references to the unique viewdatas. Hopefully that addresses some of your concerns here